src/ILCompiler.WebAssembly/src/CodeGen/EvaluationStack.cs

```diff
@@ -125,7 +125,11 @@ public void InsertAt(T v, int pos)
         /// <returns>Element at the top of the stack</returns>
         public T Peek()
         {
-            Debug.Assert(_top > 0, "Stack is not empty");
+#if DEBUG // This should eventually be an assert, but while many opcodes are unimplemented, we can just throw to avoid
```

Comment:
This comment doesn't look right: input validation should never be done with asserts. Instead of crashing the compiler, this should throw an exception that we can catch and handle,

Unless you want each callee of `Peek` to do checks whether it's safe to call `Peek`.

We have been doing `ThrowHelper.ThrowInvalidProgramException(ExceptionStringID.InvalidProgramSpecific, method);` for things like this.

Reply:
It turns out the assert comes from the cpp importer, but I'm happy to change this to throw invalid program instead. You're right that asserting is wrong here. We don't have the method in the EvaluationStack, but hopefully at least throwing the right exception here helps.